File: src/porter_lidar_processor/porter_lidar_processor/processor_node.py
"""Porter LIDAR Processor Node — subscribes to /scan, publishes /scan/processed.

Applies a configurable chain of filters to raw LaserScan data from the
ydlidar_driver, producing a cleaned scan suitable for Nav2 and obstacle
avoidance. Filters can be toggled at runtime via a ``SetBool`` service
and parameters can be changed dynamically.

Architecture:
    /scan (ydlidar_driver) → [processor_node] → /scan/processed

Filters (applied in order when enabled):
    1. Range clamp   — discard values outside [min_range, max_range]
    2. Outlier reject — remove salt-and-pepper spikes (MAD-based)
    3. Median filter  — sliding-window median for denoising
    4. Smoothing      — moving average for gentle noise reduction
    5. ROI crop       — zero out ranges outside angular region of interest
    6. Downsample     — reduce point density (optional, for RPi performance)

Copyright 2026 VirtusCo. All rights reserved. Proprietary and confidential.
"""

import numpy as np
from porter_lidar_processor.filters import (
    downsample_filter,
    median_filter,
    moving_average_filter,
    outlier_rejection_filter,
    range_clamp_filter,
    roi_crop_filter,
)
import rclpy
from rclpy.node import Node
from rclpy.qos import HistoryPolicy, QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import LaserScan
from std_srvs.srv import SetBool


class ProcessorNode(Node):
    """ROS 2 node that filters LaserScan data for Porter Robot.

    Subscribes to ``/scan`` (raw) and publishes to ``/scan/processed``
    (filtered). All filter stages can be enabled/disabled and tuned
    via ROS parameters. A ``~/enable_filters`` service toggles the
    entire filter pipeline on/off at runtime.
    """

    def __init__(self):
        """Initialise the processor node with parameters, pub/sub, and service."""
        super().__init__('lidar_processor')

        # ── Declare parameters ────────────────────────────────────────────
        self._declare_parameters()

        # ── Read initial configuration ────────────────────────────────────
        self._load_config()

        # ── QoS: match the ydlidar_driver publisher (RELIABLE + KEEP_LAST) ─
        scan_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=5,
        )

        # ── Subscriber ────────────────────────────────────────────────────
        self.scan_sub_ = self.create_subscription(
            LaserScan, 'scan', self._scan_callback, scan_qos)

        # ── Publisher ─────────────────────────────────────────────────────
        self.scan_pub_ = self.create_publisher(
            LaserScan, 'scan/processed', scan_qos)

        # ── Service: toggle filters on/off ────────────────────────────────
        self.filter_srv_ = self.create_service(
            SetBool, '~/enable_filters', self._enable_filters_callback)

        # ── Parameter change callback ─────────────────────────────────────
        self.add_on_set_parameters_callback(self._on_parameter_change)

        # ── Stats ─────────────────────────────────────────────────────────
        self.msg_count_ = 0

        self.get_logger().info(
            'Porter LIDAR Processor started — '
            f'filters_enabled={self.filters_enabled_}')

    # ══════════════════════════════════════════════════════════════════════
    # Parameter declaration
    # ══════════════════════════════════════════════════════════════════════

    def _declare_parameters(self):
        """Declare all typed parameters with defaults."""
        # Master switch
        self.declare_parameter('filters_enabled', True)

        # Range clamp
        self.declare_parameter('range_clamp.enabled', True)
        self.declare_parameter('range_clamp.min_range', 0.05)
        self.declare_parameter('range_clamp.max_range', 12.0)

        # Outlier rejection
        self.declare_parameter('outlier.enabled', True)
        self.declare_parameter('outlier.kernel_size', 5)
        self.declare_parameter('outlier.threshold', 1.5)

        # Median filter
        self.declare_parameter('median.enabled', True)
        self.declare_parameter('median.kernel_size', 5)

        # Smoothing (moving average)
        self.declare_parameter('smoothing.enabled', False)
        self.declare_parameter('smoothing.kernel_size', 3)

        # ROI crop
        self.declare_parameter('roi.enabled', False)
        self.declare_parameter('roi.angle_min_deg', -90.0)
        self.declare_parameter('roi.angle_max_deg', 90.0)

        # Downsample
        self.declare_parameter('downsample.enabled', False)
        self.declare_parameter('downsample.factor', 2)

    # ══════════════════════════════════════════════════════════════════════
    # Configuration loading
    # ══════════════════════════════════════════════════════════════════════

    def _load_config(self):
        """Read all parameters into instance variables."""
        self.filters_enabled_ = (
            self.get_parameter('filters_enabled').get_parameter_value().bool_value
        )

        # Range clamp
        self.range_clamp_enabled_ = (
            self.get_parameter('range_clamp.enabled')
            .get_parameter_value().bool_value
        )
        self.range_clamp_min_ = (
            self.get_parameter('range_clamp.min_range')
            .get_parameter_value().double_value
        )
        self.range_clamp_max_ = (
            self.get_parameter('range_clamp.max_range')
            .get_parameter_value().double_value
        )

        # Outlier rejection
        self.outlier_enabled_ = (
            self.get_parameter('outlier.enabled')
            .get_parameter_value().bool_value
        )
        self.outlier_kernel_ = (
            self.get_parameter('outlier.kernel_size')
            .get_parameter_value().integer_value
        )
        self.outlier_threshold_ = (
            self.get_parameter('outlier.threshold')
            .get_parameter_value().double_value
        )

        # Median
        self.median_enabled_ = (
            self.get_parameter('median.enabled')
            .get_parameter_value().bool_value
        )
        self.median_kernel_ = (
            self.get_parameter('median.kernel_size')
            .get_parameter_value().integer_value
        )

        # Smoothing
        self.smoothing_enabled_ = (
            self.get_parameter('smoothing.enabled')
            .get_parameter_value().bool_value
        )
        self.smoothing_kernel_ = (
            self.get_parameter('smoothing.kernel_size')
            .get_parameter_value().integer_value
        )

        # ROI crop
        self.roi_enabled_ = (
            self.get_parameter('roi.enabled')
            .get_parameter_value().bool_value
        )
        self.roi_angle_min_ = (
            self.get_parameter('roi.angle_min_deg')
            .get_parameter_value().double_value
        )
        self.roi_angle_max_ = (
            self.get_parameter('roi.angle_max_deg')
            .get_parameter_value().double_value
        )

        # Downsample
        self.downsample_enabled_ = (
            self.get_parameter('downsample.enabled')
            .get_parameter_value().bool_value
        )
        self.downsample_factor_ = (
            self.get_parameter('downsample.factor')
            .get_parameter_value().integer_value
        )

    # ══════════════════════════════════════════════════════════════════════
    # Dynamic parameter change callback
    # ══════════════════════════════════════════════════════════════════════

    def _on_parameter_change(self, params):
        """Handle dynamic parameter changes without restart.

        Args:
            params: List of changed Parameter objects.

        Returns:
            SetParametersResult indicating success.
        """
        from rcl_interfaces.msg import SetParametersResult

        for param in params:
            name = param.name
            val = param.value

            if name == 'filters_enabled':
                self.filters_enabled_ = val
                self.get_logger().info(
                    f'Filters {"enabled" if val else "disabled"}')
            elif name == 'range_clamp.enabled':
                self.range_clamp_enabled_ = val
            elif name == 'range_clamp.min_range':
                self.range_clamp_min_ = val
            elif name == 'range_clamp.max_range':
                self.range_clamp_max_ = val
            elif name == 'outlier.enabled':
                self.outlier_enabled_ = val
            elif name == 'outlier.kernel_size':
                self.outlier_kernel_ = val
            elif name == 'outlier.threshold':
                self.outlier_threshold_ = val
            elif name == 'median.enabled':
                self.median_enabled_ = val
            elif name == 'median.kernel_size':
                self.median_kernel_ = val
            elif name == 'smoothing.enabled':
                self.smoothing_enabled_ = val
            elif name == 'smoothing.kernel_size':
                self.smoothing_kernel_ = val
            elif name == 'roi.enabled':
                self.roi_enabled_ = val
            elif name == 'roi.angle_min_deg':
                self.roi_angle_min_ = val
            elif name == 'roi.angle_max_deg':
                self.roi_angle_max_ = val
            elif name == 'downsample.enabled':
                self.downsample_enabled_ = val
            elif name == 'downsample.factor':
                self.downsample_factor_ = val

        return SetParametersResult(successful=True)

    # ══════════════════════════════════════════════════════════════════════
    # Service: enable/disable filters
    # ══════════════════════════════════════════════════════════════════════

    def _enable_filters_callback(self, request, response):
        """Handle ~/enable_filters service request.

        Args:
            request: SetBool request (data=True to enable, False to disable).
            response: SetBool response.

        Returns:
            The populated response.
        """
        self.filters_enabled_ = request.data
        state = 'enabled' if request.data else 'disabled'
        response.success = True
        response.message = f'Filters {state}'
        self.get_logger().info(f'Service: filters {state}')
        return response

    # ══════════════════════════════════════════════════════════════════════
    # Scan processing callback
    # ══════════════════════════════════════════════════════════════════════

    def _scan_callback(self, msg: LaserScan):
        """Process incoming LaserScan and publish filtered result.

        Args:
            msg: Raw LaserScan from ydlidar_driver.
        """
        # Convert to numpy for efficient processing
        ranges = np.array(msg.ranges, dtype=np.float32)

        if self.filters_enabled_:
            ranges = self._apply_filters(ranges, msg)

        # Build output message — copy header and geometry from input
        out = LaserScan()
        out.header = msg.header
        out.angle_min = msg.angle_min
        out.angle_max = msg.angle_max
        out.angle_increment = msg.angle_increment
        out.time_increment = msg.time_increment
        out.scan_time = msg.scan_time
        out.range_min = msg.range_min
        out.range_max = msg.range_max
        out.ranges = ranges.tolist()

        # Copy intensities if present, otherwise empty
        if len(msg.intensities) > 0:
            out.intensities = list(msg.intensities)

        self.scan_pub_.publish(out)

        self.msg_count_ += 1
        if self.msg_count_ == 1:
            self.get_logger().info(
                f'First processed scan published '
                f'({len(ranges)} points)')

    # ══════════════════════════════════════════════════════════════════════
    # Filter pipeline
    # ══════════════════════════════════════════════════════════════════════

    def _apply_filters(
        self, ranges: np.ndarray, msg: LaserScan
    ) -> np.ndarray:
        """Apply the configured filter chain to range data.

        Filters are applied in a fixed order optimised for quality:
        1. Range clamp (remove physically impossible values first)
        2. Outlier rejection (remove spikes before smoothing)
        3. Median filter (denoise)
        4. Smoothing / moving average (optional gentle filtering)
        5. ROI crop (restrict angular field of view)
        6. Downsample (reduce density for constrained compute)

        Args:
            ranges: 1-D numpy array of range values.
            msg: Original LaserScan message (for geometry info).

        Returns:
            Filtered numpy array of ranges.
        """
        # 1. Range clamp
        if self.range_clamp_enabled_:
            ranges = range_clamp_filter(
                ranges, self.range_clamp_min_, self.range_clamp_max_)

        # 2. Outlier rejection
        if self.outlier_enabled_:
            ranges = outlier_rejection_filter(
                ranges, self.outlier_kernel_, self.outlier_threshold_)

        # 3. Median filter
        if self.median_enabled_:
            ranges = median_filter(ranges, self.median_kernel_)

        # 4. Smoothing
        if self.smoothing_enabled_:
            ranges = moving_average_filter(
                ranges, self.smoothing_kernel_)

        # 5. ROI crop
        if self.roi_enabled_:
            ranges = roi_crop_filter(
                ranges,
                msg.angle_min,
                msg.angle_max,
                msg.angle_increment,
                self.roi_angle_min_,
                self.roi_angle_max_,
            )

        # 6. Downsample
        if self.downsample_enabled_:
            ranges = downsample_filter(
                ranges, self.downsample_factor_)

        return ranges


def main(args=None):
    """Entry point for the processor node."""
    rclpy.init(args=args)
    node = ProcessorNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.get_logger().info('Shutting down Porter LIDAR Processor')
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/porter_lidar_processor_cpp/src/processor_node.cpp
// Copyright 2026 VirtusCo. All rights reserved.
// Proprietary and confidential.
//
// Implementation of the LIDAR processor node.
// Applies a 6-stage filter pipeline to raw LaserScan data:
//   1. Range clamp     — discard out-of-spec values
//   2. Outlier reject  — MAD-based spike removal
//   3. Median filter   — salt-and-pepper noise reduction
//   4. EMA smoothing   — temporal smoothing across scans
//   5. ROI crop        — angular region of interest
//   6. Downsample      — reduce point density
//
// Target: < 2ms per scan on Raspberry Pi 5.

#include "porter_lidar_processor/processor_node.hpp"
#include "porter_lidar_processor/filters.hpp"

#include <chrono>
#include <string>
#include <vector>

#include "diagnostic_msgs/msg/diagnostic_status.hpp"
#include "diagnostic_msgs/msg/key_value.hpp"

namespace porter_lidar_processor
{

ProcessorNode::ProcessorNode()
: Node("lidar_processor_node"),
  scan_count_(0),
  processing_time_ms_avg_(0.0)
{
  // ── Declare all parameters with typed defaults ──────────────────────────
  min_range_ = this->declare_parameter<double>("min_range", 0.12);
  max_range_ = this->declare_parameter<double>("max_range", 12.0);
  outlier_threshold_ = this->declare_parameter<double>("outlier_threshold", 3.0);
  median_window_ = this->declare_parameter<int>("median_window", 5);
  smoothing_alpha_ = this->declare_parameter<double>("smoothing_alpha", 0.3);
  roi_angle_min_ = this->declare_parameter<double>("roi_angle_min", -3.14159);
  roi_angle_max_ = this->declare_parameter<double>("roi_angle_max", 3.14159);
  downsample_factor_ = this->declare_parameter<int>("downsample_factor", 1);
  enable_range_clamp_ = this->declare_parameter<bool>("enable_range_clamp", true);
  enable_outlier_rejection_ = this->declare_parameter<bool>("enable_outlier_rejection", true);
  enable_median_ = this->declare_parameter<bool>("enable_median", true);
  enable_smoothing_ = this->declare_parameter<bool>("enable_smoothing", true);
  enable_roi_crop_ = this->declare_parameter<bool>("enable_roi_crop", false);
  enable_downsample_ = this->declare_parameter<bool>("enable_downsample", false);

  // ── QoS: SensorDataQoS with RELIABLE override ──────────────────────────
  // Matches ydlidar_driver publisher and RViz2 subscriber expectations.
  auto scan_qos = rclcpp::SensorDataQoS();
  scan_qos.reliable();

  // ── Subscriber ──────────────────────────────────────────────────────────
  scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
    "scan", scan_qos,
    std::bind(&ProcessorNode::on_scan, this, std::placeholders::_1));

  // ── Publishers ──────────────────────────────────────────────────────────
  scan_pub_ = this->create_publisher<sensor_msgs::msg::LaserScan>(
    "scan/processed", scan_qos);

  diag_pub_ = this->create_publisher<diagnostic_msgs::msg::DiagnosticArray>(
    "/diagnostics", 10);

  // ── Diagnostics timer: every 2 seconds ──────────────────────────────────
  diag_timer_ = this->create_wall_timer(
    std::chrono::seconds(2),
    std::bind(&ProcessorNode::publish_diagnostics, this));

  // ── Initialise last_scan_time_ ──────────────────────────────────────────
  last_scan_time_ = this->now();

  RCLCPP_INFO(
    this->get_logger(),
    "Porter LIDAR Processor (C++) started — "
    "clamp=%s outlier=%s median=%s smooth=%s roi=%s downsample=%s",
    enable_range_clamp_ ? "ON" : "OFF",
    enable_outlier_rejection_ ? "ON" : "OFF",
    enable_median_ ? "ON" : "OFF",
    enable_smoothing_ ? "ON" : "OFF",
    enable_roi_crop_ ? "ON" : "OFF",
    enable_downsample_ ? "ON" : "OFF");
}

// ═══════════════════════════════════════════════════════════════════════════════
// Scan processing callback
// ═══════════════════════════════════════════════════════════════════════════════

void ProcessorNode::on_scan(const sensor_msgs::msg::LaserScan::SharedPtr msg)
{
  const auto t_start = std::chrono::high_resolution_clock::now();

  // Work on a mutable copy of ranges
  std::vector<float> ranges(msg->ranges.begin(), msg->ranges.end());

  // ── Filter pipeline (each stage is optional) ───────────────────────────

  // 1. Range clamp — discard physically impossible values first
  if (enable_range_clamp_) {
    ranges = range_clamp(
      ranges,
      static_cast<float>(min_range_),
      static_cast<float>(max_range_));
  }

  // 2. Outlier rejection — remove spikes before smoothing
  if (enable_outlier_rejection_) {
    ranges = outlier_rejection(
      ranges,
      static_cast<float>(outlier_threshold_));
  }

  // 3. Median filter — denoise
  if (enable_median_) {
    ranges = median_filter(ranges, median_window_);
  }

  // 4. Exponential moving average smoothing — temporal consistency
  if (enable_smoothing_) {
    ranges = exponential_smoothing(
      ranges,
      previous_ranges_,
      static_cast<float>(smoothing_alpha_));
  }

  // 5. ROI crop — restrict angular field of view
  if (enable_roi_crop_) {
    ranges = roi_crop(
      ranges,
      msg->angle_min,
      msg->angle_increment,
      static_cast<float>(roi_angle_min_),
      static_cast<float>(roi_angle_max_));
  }

  // 6. Downsample — reduce point density
  if (enable_downsample_) {
    ranges = downsample(ranges, downsample_factor_);
  }

  // Store for next EMA iteration
  previous_ranges_ = ranges;

  // ── Build output message ────────────────────────────────────────────────
  auto out = sensor_msgs::msg::LaserScan();
  out.header = msg->header;
  out.angle_min = msg->angle_min;
  out.angle_max = msg->angle_max;
  out.angle_increment = msg->angle_increment;
  out.time_increment = msg->time_increment;
  out.scan_time = msg->scan_time;
  out.range_min = msg->range_min;
  out.range_max = msg->range_max;
  out.ranges = std::move(ranges);

  // Copy intensities if present
  if (!msg->intensities.empty()) {
    out.intensities = msg->intensities;
  }

  scan_pub_->publish(out);

  // ── Timing ──────────────────────────────────────────────────────────────
  const auto t_end = std::chrono::high_resolution_clock::now();
  const double elapsed_ms =
    std::chrono::duration<double, std::milli>(t_end - t_start).count();

  // Exponential moving average of processing time
  ++scan_count_;
  if (scan_count_ == 1) {
    processing_time_ms_avg_ = elapsed_ms;
  } else {
    processing_time_ms_avg_ = 0.95 * processing_time_ms_avg_ + 0.05 * elapsed_ms;
  }

  last_scan_time_ = this->now();

  // Warn if processing time exceeds 5ms (target < 2ms on RPi 5)
  if (elapsed_ms > 5.0) {
    RCLCPP_WARN(
      this->get_logger(),
      "Scan processing took %.2f ms (target < 2ms)", elapsed_ms);
  }

  // Log first scan
  if (scan_count_ == 1) {
    RCLCPP_INFO(
      this->get_logger(),
      "First processed scan published (%zu points, %.2f ms)",
      out.ranges.size(), elapsed_ms);
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Diagnostics publisher
// ═══════════════════════════════════════════════════════════════════════════════

void ProcessorNode::publish_diagnostics()
{
  auto diag_msg = diagnostic_msgs::msg::DiagnosticArray();
  diag_msg.header.stamp = this->now();

  auto status = diagnostic_msgs::msg::DiagnosticStatus();
  status.name = "porter_lidar_processor_cpp: scan_pipeline";
  status.hardware_id = "lidar_processor";

  if (scan_count_ == 0) {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
    status.message = "No scans received yet";
  } else {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
    status.message = "Filter pipeline active";
  }

  // Key-value pairs
  auto kv_count = diagnostic_msgs::msg::KeyValue();
  kv_count.key = "scans_processed";
  kv_count.value = std::to_string(scan_count_);
  status.values.push_back(kv_count);

  auto kv_time = diagnostic_msgs::msg::KeyValue();
  kv_time.key = "avg_processing_ms";
  kv_time.value = std::to_string(processing_time_ms_avg_);
  status.values.push_back(kv_time);

  // Compute scan rate from last_scan_time
  const double age_s = (this->now() - last_scan_time_).seconds();
  auto kv_age = diagnostic_msgs::msg::KeyValue();
  kv_age.key = "last_scan_age_s";
  kv_age.value = std::to_string(age_s);
  status.values.push_back(kv_age);

  // Report valid point ratio from the most recent processed scan
  if (!previous_ranges_.empty()) {
    const auto stats = compute_stats(previous_ranges_);

    auto kv_valid = diagnostic_msgs::msg::KeyValue();
    kv_valid.key = "valid_point_ratio";
    kv_valid.value = std::to_string(
      1.0f - stats.invalid_ratio);
    status.values.push_back(kv_valid);

    auto kv_total = diagnostic_msgs::msg::KeyValue();
    kv_total.key = "total_points";
    kv_total.value = std::to_string(stats.total);
    status.values.push_back(kv_total);
  }

  // Active filters summary
  std::string active_filters;
  if (enable_range_clamp_) {active_filters += "clamp ";}
  if (enable_outlier_rejection_) {active_filters += "outlier ";}
  if (enable_median_) {active_filters += "median ";}
  if (enable_smoothing_) {active_filters += "smooth ";}
  if (enable_roi_crop_) {active_filters += "roi ";}
  if (enable_downsample_) {active_filters += "downsample ";}
  if (active_filters.empty()) {active_filters = "none";}

  auto kv_filters = diagnostic_msgs::msg::KeyValue();
  kv_filters.key = "active_filters";
  kv_filters.value = active_filters;
  status.values.push_back(kv_filters);

  // Stale scan warning
  if (scan_count_ > 0 && age_s > 5.0) {
    status.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
    status.message = "No scan received for " + std::to_string(age_s) + " seconds";
  }

  diag_msg.status.push_back(status);
  diag_pub_->publish(diag_msg);
}

}  // namespace porter_lidar_processor
